### PR TITLE
Update copy_dir function to include subdirectories

### PR DESCRIPTION
**Pull request by rhodesn**
Fix for #72 so that the copy_dir function copies the source directory
tree including subdirectories. It will skip symlinks.

### CODE REVIEW

**Comment by reaperzn** [comment]
As part of this PR its probably worth adding a compare_dir function to avoid long lists of compared files to the final output.
In that compare function it would be worth only printing the specific files that do differ or if a new pre or post file exists. 

I was going to raise a separate issue for that but it would change a lot of how this PR is structured.

**Review by reaperzn — DISMISSED** [review]
see previous comment

**Comment by rhodesn** [comment]
Several updates here:
- `copy_dir` function now has a file_pattern option which allows you to filter the files from within the directory
- new `get_diff_dir` function to reduce stdout noise when diff'ing the subdirectories (can be toggled with the verbose option)

**Comment by popindavibe** [comment]
I have tested this with new or modified file in /etc/php.d and spotted few issues, mostly because the new paradigm to compare for subdirs was not applied all the way.

I have modified the code from your PR to include:
- subdirs files removal during `--overwrite` option
- checking for remaining files into subdirs during pre phases
- modified generic python error handling on missing file to prefix a '-' for pre files and a '+' for post files
- `continue` if a file is not found (instead of `return False`) 
- added a comment to find the 'MAIN' section more easily

I have done quite a few tests, but extra testing / review would be welcome. 

Here below is the patch file from my modifications:
```pr-77.patch

--- configsnap  2018-08-09 10:07:50.209740086 +0100
+++ configsnap2 2018-08-09 10:09:06.459327238 +0100
@@ -86,6 +86,11 @@
-            if any(options.phase in s for s in os.listdir(workdir)):
-                report_info(
-                    "%s files exist in %s. Overwrite (-w/--overwrite) enabled so removing." % (options.phase, workdir))
-                for reportfile in os.listdir(workdir):
-                    if options.phase in reportfile:
-                        os.remove(os.path.join(workdir, reportfile))
+            # We need to go through all subdirs as well
+            for root, dirs, files in os.walk(workdir):
+                if any(s.endswith(".%s" % options.phase) for s in files):
+                    report_info(
+                        "%s files exist in %s. Overwrite (-w/--overwrite) enabled so removing." % (options.phase, root))
+
+                # actual removal
+                for filename in files:
+                    if filename.endswith(".%s" % options.phase):
+                        os.remove(os.path.join(root, filename))
+
@@ -350,0 +356 @@
+        all_pre_files = []
@@ -351,0 +358 @@
+
@@ -352,0 +360,5 @@
+
+            # building list of all 'pre' files
+            bare_file = os.path.splitext(pre_filename)[0] + '.' + self.phase
+            all_pre_files.append(bare_file)
+
@@ -360,3 +372,2 @@
-                report_error("Unable to open file %s: %s" %
-                             (e.filename, e.strerror))
-                return False
+                report_error("- File removed: %s" % (e.filename))
+                continue
@@ -386,0 +398,7 @@
+        # Looking for new files
+        for post_filename in glob.glob(sdir + '/*' + self.phase):
+            if post_filename not in all_pre_files:
+                report_error("+ File added: %s:" % post_filename)
+
+
+
@@ -391,0 +410,8 @@
+
+
+###################################################
+#
+#                   MAIN
+#
+###################################################
+
@@ -456,5 +482,7 @@
-    for filename in os.listdir('.'):
-        if filename.endswith(".%s" % phase) and not options.compare_only:
-            report_error("Files for %s already exist in %s" % (phase,
-                                                               os.getcwd()))
-            sys.exit(1)
+    # We need to go into subdirs as well for checking
+    for root, dirs, files in os.walk(workdir):
+        for filename in files:
+            if filename.endswith(".%s" % phase) and not options.compare_only:
+                report_error("Files for %s already exist in %s" % (phase,
+                                                                   os.getcwd()))
+                sys.exit(1)
```
This can be saved as pr-77.patch and applied with:
```
patch configsnap pr-77.patch
```

**Comment by rhodesn** [comment]
@popindavibe I've applied and tested your patch and it all looks great, thanks! Pushed another update for review. (had to fix-up some PEP8 warnings regarding too many lines)

**Comment by popindavibe** [comment]
Did some new testing, works for me.

**Comment by popindavibe** [comment]
Ok, I've been looking into the additional config file for configsnap with @reaperzn , and here are my latest changes for the PR. 
**note: considering all the changes introduced, I am bumping the version number**
```patch
--- b/configsnap
+++ a/configsnap
@@ -31 +31 @@ import tarfile
-version = "0.14"
+version = "0.15"
@@ -176 +176 @@ def compare_files():
-        except Exception, e:
+        except ConfigParser.NoOptionError, e:
@@ -178,0 +179,3 @@ def compare_files():
+        except Exception, e:
+            report_error(
+                "Could not parse config file section %s: %s" % (section, e))
@@ -414 +417,4 @@ class dataGather:
-desc = 'Record useful system state information, and compare to previous state if run with PHASE containing "post" or "rollback".'
+desc = ("Record useful system state information, and compare to previous state "
+       "if run with PHASE containing \"post\" or \"rollback\".\nAn optional file, "
+       "/etc/configsnap/additional.conf, can be provided for extra files, "
+       "directories or command to register during configsnap execution.")
@@ -515 +521,2 @@ for section in Config.sections():
-        if Config.get(section, 'type').lower() != "file" and Config.get(section, 'type').lower() != "command":
+        if not re.match('^(file|directory|command)$', Config.get(section, 'type').lower()):
+        #if Config.get(section, 'type').lower() != "file" and Config.get(section, 'type').lower() != "command":
@@ -517 +524 @@ for section in Config.sections():
-                "Wrong \"Type\" defined for custom collection entry \"%s\"; should be \"file\" or \"command\"" % section)
+                "Wrong \"Type\" defined for custom collection entry \"%s\"; should be \"file\", \"directory\" or \"command\"" % section)
@@ -723 +730 @@ for section in Config.sections():
-    except Exception, e:
+    except ConfigParser.NoOptionError, e:
@@ -725,0 +733,3 @@ for section in Config.sections():
+    except Exception, e:
+        report_error(
+            "Could not parse config file section %s: %s" % (section, e))
@@ -753 +763 @@ if os.path.exists('/etc/sysconfig/network-scripts/'):
-# copy custom files
+# copy custom files and directories
@@ -755,0 +766,7 @@ for section in Config.sections():
+        # Optional options
+        if Config.has_option(section, 'failok') and Config.get(section, 'failok').lower() == "true":
+            failokcfg = True
+        # If failok is anything except true...
+        else:
+                failokcfg = False
+
@@ -759,8 +775,0 @@ for section in Config.sections():
-
-            # Optional options
-            if Config.has_option(section, 'failok') and Config.get(section, 'failok').lower() == "true":
-                failokcfg = True
-            # If failok is anything except true...
-            else:
-                failokcfg = False
-
@@ -768 +777,8 @@ for section in Config.sections():
-    except Exception, e:
+
+        elif cfgtype == 'directory':
+            cpdir = Config.get(section, 'directory')
+            if not cpdir.endswith("/"):
+                cpdir = cpdir + '/'
+            run.copy_dir(cpdir, fail_ok=failokcfg, sort=False)
+
+    except ConfigParser.NoOptionError, e:
@@ -770,0 +787,3 @@ for section in Config.sections():
+    except Exception, e:
+        report_error(
+            "Could not parse config file section %s: %s" % (section, e))
```
@nrhodes91 can you test this?

I've done a quick test with **/etc/configsnap/additional.conf**: 
```
# Optional custom command or file collection for configsnap
#
# Format:
# [section]
# Type: file
# File: [ source filename ]
#
# [section]
# Type: command
# File: [ full command to run ]
# Sort: [ True | False ] (optional)
#
# The following options apply to either "file" or "command":
# Failok: [ True | False ] - what to do if action fails (optional)
# Compare: [ True | False ] - when phase contains post or rollback, whether to diff against pre (optional)
#
#
# Examples:
# [psspecial]
# Type: command
# Command: /bin/ps -aux
#
[sshd_config]
Type: file
File: /etc/ssh/sshd_config

[ssh]
Type: directory
Directory: /etc/ssh
```

File: configsnap [hunk 1]
```diff
@@ -519,7 +519,7 @@ if os.path.exists(customcollectionfile):
 for section in Config.sections():
     if Config.has_option(section, 'type'):
         if not re.match('^(file|directory|command)$', Config.get(section, 'type').lower()):
-        #if Config.get(section, 'type').lower() != "file" and Config.get(section, 'type').lower() != "command":
+            # if Config.get(section, 'type').lower() != "file" and Config.get(section, 'type').lower() != "command":
```
Comment (popindavibe) [inline configsnap]: Sorry Nick, can we just get rid of that old line? I shouldn't have kept it in the patch.

File: configsnap [hunk 2]
```diff
@@ -415,9 +415,9 @@ class dataGather:
 ###################################################
 
 desc = ("Record useful system state information, and compare to previous state "
-       "if run with PHASE containing \"post\" or \"rollback\".\nAn optional file, "
-       "/etc/configsnap/additional.conf, can be provided for extra files, "
-       "directories or command to register during configsnap execution.")
+        "if run with PHASE containing \"post\" or \"rollback\".\nAn optional file, "
+        "/etc/configsnap/additional.conf, can be provided for extra files, "
+        "directories or command to register during configsnap execution.")
```
Comment (popindavibe) [inline configsnap]: command**s** (my fault, sorry!)

**Comment by popindavibe** [comment]
Added few comments (for typos or useless commented out lines that I kept in the proposed patch). Looks good though, almost there!

**Review by popindavibe — CHANGES_REQUESTED** [review]
Added few comments (for typos or useless commented out lines that I kept in the proposed patch). Looks good though, almost there!

**Review by popindavibe — APPROVED** [review]

**Comment by popindavibe** [comment]
Found a discrepancy for the output result:
```
Starting sub-dir diffs...
No differences found for files in /home/jean/testing-pr77/configsnap/network-scripts
No differences found for files in /home/jean/testing-pr77/configsnap/php.d
+ File added: /home/jean/testing-pr77/configsnap/php.d/curl.ini.test4:
+ File added: /home/jean/testing-pr77/configsnap/php.d/fileinfo.ini.test4:
+ File added: /home/jean/testing-pr77/configsnap/php.d/json.ini.test4:
+ File added: /home/jean/testing-pr77/configsnap/php.d/phar.ini.test4:
+ File added: /home/jean/testing-pr77/configsnap/php.d/zip.ini.test4:
```

```patch
[root@c6 configsnap]# git diff -U0 56e598bdc665f1ce8e35c3e0811d6ad369f827c6 configsnap
diff --git a/configsnap b/configsnap
index 8af1fc3..970256d 100755
--- a/configsnap
+++ b/configsnap
@@ -398,3 +397,0 @@ class dataGather:
-        if diffs_found_msg is False:
-            report_info("No differences found for files in %s" % sdir)
-
@@ -404,0 +402,5 @@ class dataGather:
+                diffs_found_msg = True
+
+        if diffs_found_msg is False:
+            report_info("No differences found for files in %s" % sdir)
+
```
gives
```
Starting sub-dir diffs...
No differences found for files in /home/jean/testing-pr77/configsnap/network-scripts
+ File added: /home/jean/testing-pr77/configsnap/php.d/curl.ini.test4:
+ File added: /home/jean/testing-pr77/configsnap/php.d/fileinfo.ini.test4:
+ File added: /home/jean/testing-pr77/configsnap/php.d/json.ini.test4:
+ File added: /home/jean/testing-pr77/configsnap/php.d/phar.ini.test4:
+ File added: /home/jean/testing-pr77/configsnap/php.d/zip.ini.test4:
```

Testing pattern now.

**Comment by popindavibe** [comment]
Tested with multiple scenarios, including the new File_Pattern options in `/etc/configsnap/additional.conf`.

Approved

**Pull request merged**